### ISSUE TITLE
Use smartCase for Find

### ISSUE THREAD

**Issue by johnbillion** [Microsoft/vscode#55247]
Issue Type: <b>Feature Request</b>

The `search.smartCase` setting is really useful. Thanks for implementing it.

It would be great if it worked in the "Find" functionality too, so I can quickly perform a case sensitive search within the text in a file.

VS Code version: Code 1.25.1 (1dfc5e557209371715f655691b1235b6b26a06be, 2018-07-11T15:33:29.235Z)
OS version: Darwin x64 16.7.0

**Comment by sandyarmstrong** [comment]
Yes, and it would be nice to adopt [this UI improvement](https://github.com/microsoft/vscode/issues/42687) to let the user know when they are in smart case mode.

**Comment by bogem** [comment]
Any updates?

**Comment by vegerot** [comment]
What's the news?

**Comment by vegerot** [comment]
Is this feature up for grabs?  Is anybody working on this rn?

**Comment by ISAACRITHARSON** [comment]
# RFC: Smart Case for the editor Find/Replace widget

Tracks: [microsoft/vscode#55247](https://github.com/microsoft/vscode/issues/55247) ("Use smartCase for Find") — community-upvoted past the maintainers' backlog threshold, still open, unimplemented, no linked PR.

## Problem

The global **Search** panel already supports `search.smartCase`: when enabled, a search is case-insensitive unless the query itself contains an uppercase letter, in which case it becomes case-sensitive automatically (the same "smart case" behavior found in Vim, IntelliJ, etc.). The in-editor **Find widget** (`Ctrl+F` / `Cmd+F`) has no equivalent — its "Match Case" toggle is a plain boolean the user must flip by hand. Users who rely on smart case in the Search panel get inconsistent behavior the moment they switch to in-editor Find.

## Proposed solution

Add a new opt-in setting, `editor.find.smartCase` (boolean, default `false`), following the naming and default-off convention of the other `editor.find.*` toggles (`globalFindClipboard`, `addExtraSpaceOnTop`, `closeOnResult`, etc. — see `src/vs/editor/common/config/editorOptions.ts:1800-1870`). When enabled:

- If the "Match Case" toggle is off and the search string contains an uppercase letter, the search behaves as case-sensitive.
- If the search string is all lowercase, or the user has explicitly turned "Match Case" on, behavior is unchanged.

This mirrors `search.smartCase`'s semantics exactly, so the mental model transfers between the Search panel and the Find widget.

## Code design

**Reuse, don't reinvent.** The global Search feature already solves this in `src/vs/workbench/services/search/common/queryBuilder.ts:322-335`:

```ts
private isCaseSensitive(contentPattern: IPatternInfo, options: ITextQueryBuilderOptions): boolean {
	if (options.isSmartCase) {
		if (contentPattern.isRegExp) {
			if (strings.containsUppercaseCharacter(contentPattern.pattern, true)) {
				return true;
			}
		} else if (strings.containsUppercaseCharacter(contentPattern.pattern)) {
			return true;
		}
	}
	return !!contentPattern.isCaseSensitive;
}
```

`strings.containsUppercaseCharacter` (`src/vs/base/common/strings.ts:947`) already takes an `ignoreEscapedChars` flag for regex patterns (so an escaped `\U` doesn't falsely count as "contains uppercase"). The editor's Find feature can depend on `base/common` directly, so this helper can be reused as-is with zero duplication.

**Where it plugs into the Find widget:**

1. **Setting** — add `editor.find.smartCase` next to the other `editor.find.*` options in `editorOptions.ts`, in the same schema block and in the `IEditorFindOptions` type. Default `false`, so existing behavior for every current user is unchanged.

2. **Single resolution point** — `FindModelBoundToEditorModel` (`src/vs/editor/contrib/find/browser/findModel.ts`) is the one class that turns `FindReplaceState` into actual searches. Today it reads `this._state.matchCase` directly at five call sites (`research()`'s `_findMatches`, `_prevSearchPosition`/`_nextSearchPosition`'s `findPreviousMatch`/`findNextMatch`, and the `SearchParams` construction around line 533). Add one private helper:

   ```ts
   private _getEffectiveMatchCase(): boolean {
   	if (this._state.matchCase) {
   		return true;
   	}
   	if (!this._editor.getOption(EditorOption.find).smartCase) {
   		return false;
   	}
   	return this._state.isRegex
   		? containsUppercaseCharacter(this._state.searchString, true)
   		: containsUppercaseCharacter(this._state.searchString);
   }
   ```

   Then swap the five `this._state.matchCase` reads inside `FindModelBoundToEditorModel` for `this._getEffectiveMatchCase()`.

3. **Everything else is untouched by design.** `FindReplaceState.matchCase` itself, the "Match Case" checkbox in `findWidget.ts`, `preserveCase`/replace logic, and the `matchCaseOverride` one-shot mechanism (used by `Find with Args`) all keep their current meaning. Smart case only changes what *effective* case-sensitivity the model computes internally — it never changes what the toggle displays or what gets persisted to find history. This keeps the change surface to one new setting + one new private method + five one-line call-site swaps, all inside a single file.

## Alternatives considered

- **Route through `matchCaseOverride` (`FindOptionOverride`)** — rejected. That mechanism is a one-shot, per-invocation override used by `editor.action.findWithArgs`, not a persistent, per-keystroke derived value. Reusing it for smart case would conflate two different lifecycles and complicate `FindReplaceState`'s existing tri-state override logic for no benefit.
- **Make smart case the default (always on)** — rejected. It would silently change search behavior for every existing user, which is surprising and inconsistent with how every other `editor.find.*` behavior toggle in this codebase defaults to off.
- **Add a visual indicator when smart case silently makes a search case-sensitive** (requested in a comment on a related issue, #42687) — deferred as a non-goal for this change; it's a separate, independent UI enhancement that can be proposed on top of this later without changing the design above.

## Non-goals

- No changes to the Replace field's `preserveCase` behavior.
- No new keybinding or toggle button — this is a settings-only opt-in.
- No visual "smart case is active" indicator (see Alternatives).

## Test plan

Add cases to `src/vs/editor/contrib/find/test/browser/findModel.test.ts` covering:
- `smartCase: false` (default) — behavior identical to today, for both lowercase and mixed-case search strings.
- `smartCase: true` + all-lowercase search string — matches are found case-insensitively.
- `smartCase: true` + a search string containing an uppercase letter — matches are found case-sensitively.
- `smartCase: true` + `matchCase` explicitly toggled on — case-sensitive regardless of pattern content (no regression from the explicit toggle).
- `smartCase: true` + regex mode with an uppercase letter inside an escape sequence (e.g. `\U`) — not treated as case-sensitive, matching the `ignoreEscapedChars` behavior already used by `search.smartCase`.

## Risk / rollout

Zero risk to existing users: the setting defaults to `false`, and when disabled the new helper is behaviorally identical to the current direct reads of `this._state.matchCase`. The diff is contained to one new setting definition and one file (`findModel.ts`) for the core logic, plus its test file.
